src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs

```diff
@@ -0,0 +1,306 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.NET.Sdk.WorkloadManifestReader
+{
+    /// <remarks>
+    /// This very specifically exposes only the functionality needed right now by the MSBuild workload resolver
+    /// and by the template engine. More general APIs will be added later.
+    /// </remarks>
+    public class WorkloadResolver : IWorkloadResolver
+    {
+        private readonly Dictionary<WorkloadDefinitionkId, WorkloadDefinition> _workloads = new Dictionary<WorkloadDefinitionkId, WorkloadDefinition>();
+        private readonly Dictionary<WorkloadPackId, WorkloadPack> _packs = new Dictionary<WorkloadPackId, WorkloadPack>();
+        private string[] _platformIds;
+        private readonly string _dotNetRootPath;
+
+        private Func<string, bool>? _fileExistOverride;
+        private Func<string, bool>? _directoryExistOverride;
+
+        public WorkloadResolver(IWorkloadManifestProvider manifestProvider, string dotNetRootPath)
+        {
+            this._dotNetRootPath = dotNetRootPath;
+
+            // eventually we may want a series of fallbacks here, as rids have in general
+            // but for now, keep it simple
+            var platformId = GetHostPlatformId();
+            if (platformId != null)
+            {
+                _platformIds = new[] { platformId, "*" };
+            }
+            else
+            {
+                _platformIds = new[] { "*" };
+            }
+
+            var manifests = new List<WorkloadManifest>();
+
+            foreach (var manifestStream in manifestProvider.GetManifests())
+            {
+                using (manifestStream)
+                {
+                    var manifest = WorkloadManifestReader.ReadWorkloadManifest(manifestStream);
+                    manifests.Add(manifest);
+                }
+            }
+
+            foreach (var manifest in manifests)
+            {
+                foreach (var workload in manifest.Workloads)
+                {
+                    _workloads.Add(workload.Key, workload.Value);
```

Comment:
Will this throw if duplicate workloads or packs are defined in different workload manifests?  Should we have better error handling?

Reply:
Good catch. This should still be an error IMO - right now, the approach is to fail on validation errors, in absence of a well-defined recovery strategy, but this could absolutely be a better error.